etcd/src/main/scala/io/buoyant/etcd/Key.scala

```diff
@@ -0,0 +1,282 @@
+      falseParam("prevExist", prevExist)
+
+    val req = mkReq(uriPath, Method.Put, params)
+    client(req).flatMap { rsp => Future.const(NodeOp.mk(req, rsp, key, params)) }
+  }
+
+  /**
+   * Delete a node.
+   *
+   * If `dir` is not true and the key is a directory, this will
+   * operation fail.
+   *
+   * If `dir` and `recursive` are true, the entire tree is deleted.
+   */
+  def delete(
+    dir: Boolean = false,
+    recursive: Boolean = false
+  ): Future[NodeOp] = {
+    val params = Params ++
+      trueParam("dir", dir) ++
+      trueParam("recursive", recursive)
+    val req = mkReq(uriPath, Method.Delete, params)
+    client(req).flatMap { rsp => Future.const(NodeOp.mk(req, rsp, key, params)) }
+  }
+
+  /** Find the greatest index referenced in a tree of nodes. */
+  private[this] def getIndex(node: Node): Long = node match {
+    case Node.Data(_, idx, _, _, _) => idx
+    case Node.Dir(_, idx, _, _, nodes) => nodes.foldLeft(idx)(_ max getIndex(_))
+  }
+
+  /**
+   * An Event constructed by watching an etcd key.
+   *
+   * If `recursive` is true, the key's subtree is observed for
+   * changes.
+   *
+   * When an unexpected error is encountered communicating with the
+   * API, the failure is published on the Event and the `backoff`
+   * stream is used to compute the time to wait before retrying. If
+   * the `backoff` stream is exhausted or a fatal error is
+   * encountered, it is reported and polling stops.
+   *
+   * The Event is not reference-counted, so each observer initiates
+   * its own polling loop. This ensures that the initial state of a
+   * tree is reported properly.
+   */
+  def events(
+    recursive: Boolean = false,
+    backoff: Stream[Duration] = Stream.empty
+  ): Event[Try[NodeOp]] = new Event[Try[NodeOp]] {
+    private[this] val origBackoff = backoff
+
+    def register(witness: Witness[Try[NodeOp]]) = {
+      @volatile var closed = false
+
+      def loop(idx: Option[Long], backoff: Stream[Duration]): Future[Unit] =
+        if (!closed) {
+          get(recursive, wait = idx.isDefined, waitIndex = idx).transform {
+            case note@Return(op) =>
+              witness.notify(note)
+              loop(Some(getIndex(op.node) + 1), origBackoff)
+
+            case note@Throw(ApiError(ApiError.KeyNotFound, _, _, idx)) =>
+              witness.notify(note)
+              loop(Some(idx + 1), origBackoff)
+
+            case note@Throw(ApiError(ApiError.EventIndexCleared, _, _, idx)) =>
+              loop(Some(idx + 1), origBackoff)
+
+            case note@Throw(NonFatal(e)) =>
+              backoff match {
+                case wait #:: backoff =>
+                  witness.notify(note)
+                  loop(None, backoff)
+
+                case _ =>
+                  witness.notify(Throw(BackoffsExhausted(key, e)))
+                  Future.Unit
+              }
+
+            case note@Throw(_) =>
+              witness.notify(note)
```

Comment:
thought was that the consumer of this can log... (we don't have a logger in this module as it is)